src/USER-PLUMED/fix_plumed.cpp

```diff
@@ -0,0 +1,436 @@
+  //   whereas if partitions are not defined then world is equal to MPI_COMM_WORLD.
+  p->cmd("setMPIComm",&world);
+
+// Set up units
+// LAMMPS units wrt kj/mol - nm - ps
+// Set up units
+
+  if (force->boltz == 1.0){
+// LAMMPS units lj
+    p->cmd("setNaturalUnits");
+  } else {
+    double energyUnits=1.0;
+    double lengthUnits=1.0;
+    double timeUnits=1.0;
+    if (force->boltz == 0.0019872067){
+// LAMMPS units real :: kcal/mol; angstrom; fs
+      energyUnits=4.184;
+      lengthUnits=0.1;
+      timeUnits=0.001;
+    } else if (force->boltz == 8.617343e-5){
+// LAMMPS units metal :: eV; angstrom; ps
+      energyUnits=96.48530749925792;
+      lengthUnits=0.1;
+      timeUnits=1.0;
+    } else if (force->boltz == 1.3806504e-23){
+// LAMMPS units si :: Joule, m; s
+      energyUnits=0.001;
+      lengthUnits=1.e-9;
+      timeUnits=1.e-12;
+    } else if (force->boltz == 1.3806504e-16){
+// LAMMPS units cgs :: erg; cms;, s
+      energyUnits=6.0221418e13;
+      lengthUnits=1.e-7;
+      timeUnits=1.e-12;
+    } else if (force->boltz == 3.16681534e-6){
+// LAMMPS units electron :: Hartree, bohr, fs
+      energyUnits=2625.5257;
+      lengthUnits=0.052917725;
+      timeUnits=0.001;
+    } else error->all(FLERR,"Odd LAMMPS units, plumed cannot work with that");
+    p->cmd("setMDEnergyUnits",&energyUnits);
+    p->cmd("setMDLengthUnits",&lengthUnits);
+    p->cmd("setMDTimeUnits",&timeUnits);
+  }
+
+// Read fix parameters:
+  int next=0;
+  for(int i=3;i<narg;++i){
+    if(!strcmp(arg[i],"outfile")) next=1;
+    else if(next==1){
+      if(universe->existflag == 1){
+        // Each replica writes an independent log file
+        //  with suffix equal to the replica id
+        char str_num[32], logFile[1024];
+        sprintf(str_num,".%d",universe->iworld);
+        strncpy(logFile,arg[i],1024-32);
+        strcat(logFile,str_num);
+        p->cmd("setLogFile",logFile);
+        next=0;
+      } else {
+        // partition option not used
+        p->cmd("setLogFile",arg[i]);
+        next=0;
+      }
+    }
+    else if(!strcmp(arg[i],"plumedfile"))next=2;
+    else if(next==2){
+      p->cmd("setPlumedDat",arg[i]);
+      next=0;
+    }
+    else error->all(FLERR,"syntax error in fix plumed - use 'fix name plumed plumedfile plumed.dat outfile plumed.out' ");
+  }
+  if(next==1) error->all(FLERR,"missing argument for outfile option");
+  if(next==2) error->all(FLERR,"missing argument for plumedfile option");
+
+  p->cmd("setMDEngine","LAMMPS");
+
+  int natoms=int(atom->natoms);
+  p->cmd("setNatoms",&natoms);
+
+  double dt=update->dt;
+  p->cmd("setTimestep",&dt);
+
+  virial_flag=1;
+  thermo_virial=1;
+  scalar_flag = 1;
+
+// This is the real initialization:
+  p->cmd("init");
+
+// Define compute to calculate potential energy
+  id_pe = new char[7];
+  id_pe = (char *) "plmd_pe";
+  char **newarg = new char*[3];
+  newarg[0] = id_pe;
+  newarg[1] = (char *) "all";
```

Comment:
Please note that fixes in LAMMPS have to provide a group id, so they can be (optionally) applied to a specific group only. now you can decide, that only group "all" is supported and the group id is ignored. But it may still be, that you set up a system, where atoms are not supposed to contribute to the total potential energy or (global) presssure.

So there should be support for the `Fix::modify_params()` API, so that you can set which PE compute, or pressure compute etc. are supposed to be used. also, it is good practice. to derive the fix or compute ID for internal fixes or computes by appending some string in capital letters to the fix id of the fix. but not use a fixed name.

If you only want to support a single instance of fix plumed, you should add a test for that, too.

Reply:
@akohlmey thank you for your comments. I think it is better to raise an error if a group id different from "all" is used. Otherwise users might think that plumed is affecting only the atoms in the group id. @gtribello do you agree?

It would be interesting to add the possibility of biasing the energy of a group of atoms but it would involve changing a large part of the code.

Reply:
@PabloPiaggi @gtribello a hard error is quite drastic. A warning message that the group is ignored, if it is not all, might be sufficient. It is more important to ensure, that the fix is not instantiated twice.

There are several suggestions for changes now pending. I can help with some of them. So please let me know what you are working on first and where you would like to have help or explanations from me.

Reply:
a test for the fix group and for creating fix plumed multiple times has been included. the former will raise a warning and explain that the group will be ignored, the latter a hard error.